src/internals/ir/callGraph.ts

```diff
@@ -431,3 +494,78 @@ export class CallGraph {
 export function isSelf(expr: AstExpression): boolean {
   return expr.kind === "id" && (expr as AstId).text === "self";
 }
+
+/**
+ * Helper function to determine if an expression is a contract state read.
+ * @param expr The expression to check.
+ * @returns True if the expression reads from a state variable; otherwise, false.
+ */
+function isContractStateRead(expr: AstExpression): boolean {
+  if (expr.kind === "field_access") {
+    const fieldAccess = expr as AstFieldAccess;
+    if (fieldAccess.aggregate.kind === "id") {
+      const idExpr = fieldAccess.aggregate as AstId;
+      if (idExpr.text === "self") {
+        return true; // Accessing a state variable via 'self'
+      }
+    }
+  }
+  return false;
+}
+
+/**
+ * Helper function to determine if a statement is a contract state write.
+ * @param stmt The statement to check.
+ * @returns True if the statement writes to a state variable; otherwise, false.
+ */
+function isContractStateWrite(
+  stmt: AstStatementAssign | AstStatementAugmentedAssign,
+): boolean {
+  const pathExpr = stmt.path;
+  if (pathExpr.kind === "field_access") {
+    const fieldAccess = pathExpr as AstFieldAccess;
+    if (fieldAccess.aggregate.kind === "id") {
+      const idExpr = fieldAccess.aggregate as AstId;
+      if (idExpr.text === "self") {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/**
+ * Helper function to determine if an expression is a blockchain state read.
+ * @param expr The expression to check.
+ * @returns True if the expression reads blockchain state; otherwise, false.
+ */
+function isBlockchainStateRead(expr: AstExpression): boolean {
+  if (expr.kind === "static_call") {
+    const staticCall = expr as AstStaticCall;
+    const functionName = staticCall.function?.text;
+    return functionName === "now" || functionName === "timestamp";
+  } else if (expr.kind === "method_call") {
+    const methodCall = expr as AstMethodCall;
+    const methodName = methodCall.method?.text;
+    return methodName === "now" || methodName === "timestamp";
+  }
+  return false;
+}
+
+/**
+ * Helper function to determine if an expression is a randomness call.
+ * @param expr The expression to check.
+ * @returns True if the expression introduces randomness; otherwise, false.
+ */
+function isRandomnessCall(expr: AstExpression): boolean {
+  if (expr.kind === "static_call") {
+    const staticCall = expr as AstStaticCall;
+    const functionName = staticCall.function?.text;
+    return functionName === "random";
```

Comment:
There are a couple of random generators: https://github.com/nowarp/misti/blob/b106ae04ee057b0650c98e037aa31efcad3d0310/src/detectors/builtin/ensurePrgSeed.ts#L46